rust/rope/src/tree.rs

```diff
@@ -566,33 +581,76 @@ impl<'a, N: NodeInfo> Cursor<'a, N> {
+    /// discontinuity.
     pub fn next<M: Metric<N>>(&mut self) -> Option<(usize)> {
         if self.position >= self.root.len() || self.leaf.is_none() {
             self.leaf = None;
             return None;
         }
-        // TODO: walk up tree to skip measure-0 nodes
-        loop {
-            if let Some(l) = self.leaf {
-                let offset_in_leaf = self.position - self.offset_of_leaf;
-                if let Some(offset_in_leaf) = M::next(l, offset_in_leaf) {
-                    if offset_in_leaf == l.len() &&
-                            self.offset_of_leaf + offset_in_leaf != self.root.len() {
-                        let _ = self.next_leaf();
-                    } else {
-                        self.position = self.offset_of_leaf + offset_in_leaf;
+
+        if let Some(offset) = self.next_inside_leaf::<M>() {
+            return Some(offset);
+        }
+
+        if let Some(l) = self.leaf {
+            self.position = self.offset_of_leaf + l.len();
+            for i in 0..CURSOR_CACHE_SIZE {
+                if self.cache[i].is_none() {
+                    // we are at the root of the tree.
+                    return Some(self.root.len());
+                }
+                let (node, j) = self.cache[i].unwrap();
+                let (next_j, offset) = node.next_positive_measure_child::<M>(j+1);
+                self.position += offset;
+                if let Some(next_j) = next_j {
+                    self.cache[i] = Some((node, next_j));
+                    let mut node_down = &node.get_children()[next_j];
+                    for k in (0..i).rev() {
+                        let (pm_child, offset) = node_down.next_positive_measure_child::<M>(0);
+                        let pm_child = pm_child.unwrap(); // at least one child must have positive measure
+                        self.position += offset;
+                        self.cache[k] = Some((node_down, pm_child));
+                        node_down = &node_down.get_children()[pm_child];
                     }
-                    return Some(self.position);
+                    self.leaf = Some(node_down.get_leaf());
+                    self.offset_of_leaf = self.position;
+                    return self.next_inside_leaf::<M>();
                 }
-                if self.offset_of_leaf + l.len() == self.root.len() {
-                    self.position = self.root.len();
-                    return Some(self.position);
+            }
+            // At this point, we know that (1) the next boundary is not not in
+            // the cached subtree, (2) self.position corresponds to the begining
+            // of the first leaf after the cached subtree.
+            self.descend();
+            return self.next::<M>();
+        } else {
+            panic!("inconsistent, shouldn't get here");
+        }
+    }
+
+    /// Tries to find the next boundary in the leaf the cursor is currently in.
+    #[inline(always)]
```

Comment:
I tried, and `#[inline]` does not seem to have an effect. I think it's because the function is too big for the LLVM heuristic, but here inlining can remove at least one branch and a few loads.